### PR TITLE
Allow use of CIDR ranges in maintenance allow list

### LINKED ISSUE #233: IP range for maintenance mode
### Description (*)

When we add Magento in maintenance mode and declare IP address exemptions it would be useful if we could add also IP address host range or netmask. This would be needed specially for IPv6. 

### Expected behavior (*)

We could add for example;
bin/magento maintenance:allow-ips xxx.xxx.xxx.xxx/28   (IPv4)
bin/magento maintenance:allow-ips xxxx:xxxx:xxxx:xxxx:0:0:0:0/56   (IPv6)

### Benefits

Now days you get from your ISP (although that you have a static IPv6) a range of IPv6 addresses that are then dynamically assigned to users in particular network.

### LINKED ISSUE #37943: [Issue] Allow use of CIDR ranges in maintenance allow list
This issue is automatically created based on existing pull request: magento/magento2#30699: Allow use of CIDR ranges in maintenance allow list

---------
### Description
The current maintenance mode allow IP list only permits individual addresses to be allowed to bypass maintenance mode. This pull request expands this to allow ranges in CIDR format.

### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/407

### Fixed Issues
1. Fixes https://github.com/magento/community-features/issues/233

### Manual testing scenarios
1. Enable maintenance mode with `php bin/magento maintenance:enable`
1. Allow individual IP address, eg `php bin/magento maintenance:allow-ips --add 192.168.10.180`
1. Allow multiple individual IP addresses, eg `php bin/magento maintenance:allow-ips --add 192.168.10.170 192.168.10.190`
1. Allow range of IP addresses, eg `php bin/magento maintenance:allow-ips --add 192.168.10.160/27`
1. Allow IPv4 & IPv6 addresses, eg `php bin/magento maintenance:allow-ips --add 192.168.46.0/24 fd12:3456:789a:1::1/64`

### Questions or comments
*None*

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by fredden**
### Description
The current maintenance mode allow IP list only permits individual addresses to be allowed to bypass maintenance mode. This pull request expands this to allow ranges in CIDR format.

### Fixed Issues
1. Fixes https://github.com/magento/community-features/issues/233

### Manual testing scenarios
1. Enable maintenance mode with `php bin/magento maintenance:enable`
1. Allow individual IP address, eg `php bin/magento maintenance:allow-ips --add 192.168.10.180`
1. Allow multiple individual IP addresses, eg `php bin/magento maintenance:allow-ips --add 192.168.10.170 192.168.10.190`
1. Allow range of IP addresses, eg `php bin/magento maintenance:allow-ips --add 192.168.10.160/27`
1. Allow IPv4 & IPv6 addresses, eg `php bin/magento maintenance:allow-ips --add 192.168.46.0/24 fd12:3456:789a:1::1/64`

### Questions or comments
*None*

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#37943: Allow use of CIDR ranges in maintenance allow list

### CODE REVIEW

**Comment by fredden** [comment]
@magento run all tests

File: composer.json
Comment (ihor-sviziev) [inline composer.json]: Seems like should be also added to the framework and setup composer.json files

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi,

Could you update your PR with one suggestion, fix failing tests and cover your changes with tests?

**Comment by okorshenko** [comment]
@maget0 import pull request to https://github.com/magento-tsg/magento2ce

**Comment by okorshenko** [comment]
@magento import pull request to https://github.com/magento-tsg/magento2ce

**Comment by magento-engcom-team** [comment]
@okorshenko the pull request successfully imported.

File: lib/internal/Magento/Framework/App/Test/Unit/MaintenanceModeTest.php
Comment (fredden) [inline lib/internal/Magento/Framework/App/Test/Unit/MaintenanceModeTest.php:108]: I'm not sure if it's correct to change this, or if I should make the class behave like the old test expects.

The old test expects: when maintenance mode is enabled, and an allow-list file exists but is empty, then maintenance mode is off.

The new behaviour here feels like an improvement on this, but I might be missing something.

**Comment by fredden** [comment]
@magento run all tests

File: composer.json
Comment (fredden) [reply composer.json]: I've added this requirement to `lib/internal/Magento/Framework/composer.json`, but didn't see any `composer.json` files within `setup/`.

File: lib/internal/Magento/Framework/App/Test/Unit/MaintenanceModeTest.php
Comment (ihor-sviziev) [reply lib/internal/Magento/Framework/App/Test/Unit/MaintenanceModeTest.php:108]: Looks like now it's more correct. I would keep the new behavior

File: lib/internal/Magento/Framework/App/Test/Unit/MaintenanceModeTest.php
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/App/Test/Unit/MaintenanceModeTest.php:242]: I believe we still should support regular IP address, just 198.51.100.1.

Could you add back such use case?

File: lib/internal/Magento/Framework/App/MaintenanceMode.php
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/App/MaintenanceMode.php]: This is quite complex method. Could you extract it to separate class and cover this logic with unit tests?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by fredden** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/App/Test/Unit/MaintenanceModeTest.php
Comment (fredden) [reply lib/internal/Magento/Framework/App/Test/Unit/MaintenanceModeTest.php:242]: I've added a test to ensure that raw IP addresses (old/current file format) still work.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @fredden ,
Could you review and fix failing tests?

**Comment by fredden** [comment]
* Integration Tests are marked here as failed, but all the detail pages show green with no amber nor red.
* Static Tests & Unit Tests are failing due to a problems with the testing framework. How do I get the testing framework to read `composer.json`?

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@engcom-Charlie could you help with failing unit & static tests? Seems like there some additional changes might be required for the b2b/ee versions.

File: lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php]: I believe we shouldn't use reference here, it's not really logical to modify original array when you're still returning result array

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php
Comment (fredden) [reply lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php]: Good catch. This is left over from an architecture change I made. I'll fix this up now.

**Comment by fredden** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Database Compare, Functional Tests B2B, Functional Tests EE

**Comment by fredden** [comment]
I've opened a pull request for Commerce and mentioned this in the "Related Pull Requests" section of both descriptions. This has worked on another pull request where changes to both were required, but isn't working here.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@fredden thx, usually we should rerun all tests after adding related pull request

**Comment by ihor-sviziev** [comment]
@fredden seems like you created or to wrong repo, you can find all used repos on the following page https://github.com/magento/magento2/runs/1367438381

**Comment by fredden** [comment]
I did run all tests. Please see this pull request where pull requests from exactly these same two repositories are linked correctly: https://github.com/magento/magento2/pull/28216/checks?check_run_id=1197882558
What's different / wrong here?

**Comment by ihor-sviziev** [comment]
@fredden unfortunately I don't know. @sidolov @gabrieldagama could you help with linking related pull request?

**Comment by sidolov** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8444](https://jira.corp.magento.com/browse/ENGCOM-8444) has been created to process this Pull Request

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @fredden,
Could you resolve conflicts with 2.4-develop branch?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Integration Tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@sidolov @sivaschenko @gabrieldagama could you provide info why tests are failing without build reports?

**Comment by gabrieldagama** [comment]
Hi @ihor-sviziev, @fredden 

There are conflicts on the related PR magento/partners-magento2ee#407, the conflicts have to be resolved before running the tests.

**Comment by fredden** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE

File: lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php]: can we use === here?

File: app/code/Magento/Backend/Model/Validator/IpValidator.php
Comment (ihor-sviziev) [inline app/code/Magento/Backend/Model/Validator/IpValidator.php]: Seems like dependency should be added to module's composer.json file

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php
Comment (fredden) [reply lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php]: We can; should we?

File: app/code/Magento/Backend/Model/Validator/IpValidator.php
Comment (fredden) [reply app/code/Magento/Backend/Model/Validator/IpValidator.php]: Sorry, I saw the files were moved in the most recent merge, but forgot to add the dependency here. I'll get that sorted.

File: lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php
Comment (fredden) [reply lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php]: Actually, no we can't. These functions both return an IP object. We could cast them to strings and strictly compare those, but I think the current implementation is more legible than that would be.

Comment (ihor-sviziev) [reply lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php]: Usually using == instead of === in some cases causing not obvious issues in not obvious cases. I really prefer to have strict comparison here, even if we'll add some string comparison

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php
Comment (fredden) [reply lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php]: Ihor, is this what you're looking for?
Screenshot_2020-11-26_21-59-20

**Comment by fredden** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php
Comment (ihor-sviziev) [reply lib/internal/Magento/Framework/App/Utility/IPAddressNormaliser.php]: yeah! something like that!

**Comment by fredden** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]
Approved ✔

Unfortunately we have MHI failure due to using static method of some library that will require some approval, but hope it will be fast

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8444](https://jira.corp.magento.com/browse/ENGCOM-8444) has been created to process this Pull Request

**Comment by fredden** [comment]
@magento run all tests

**Comment by PascalBrouwers** [comment]
@ihor-sviziev this looks good. Can you update it for the latest development version so there is no merge conflict?

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by MartynasKasp** [comment]
@fredden any updates on this? It would be very useful to have this feature.

**Comment by fredden** [comment]
@MartynasKasp what would you like from me? I keep resolving the merge conflicts, but am otherwise waiting for Adobe to process this pull request.

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
Hi @fredden ,

Thanks for your contribution and collaboration.

After taking PR changes I am getting an error in the terminal as shown in the screenshot:

Screenshot 2023-08-30 at 3 57 03 PM

Below are the commands that I ran in the terminal:
- Allow a range of IP addresses, e.g. `php bin/magento maintenance:allow-ips --add 192.168.10.160/27`
- Allow IPv4 & IPv6 addresses, e.g. `php bin/magento maintenance:allow-ips --add 192.168.46.0/24 fd12:3456:789a:1::1/64`

Could you please give me more information on whether I am missing anything?

Thanks

**Comment by fredden** [comment]
@engcom-Lima please run `composer install`.

**Comment by engcom-Lima** [comment]
### :heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento **`2.4-develop`** and **`PHP 8.1`**

**Manual testing scenario:**
- Enable maintenance mode with `php bin/magento maintenance:enable`
- Allow individual IP address, eg `php bin/magento maintenance:allow-ips --add 192.168.10.180`
- Allow multiple individual IP addresses, eg `php bin/magento maintenance:allow-ips --add 192.168.10.170 192.168.10.190`
- Allow a range of IP addresses, eg `php bin/magento maintenance:allow-ips --add 192.168.10.160/27`
- Allow IPv4 & IPv6 addresses, eg `php bin/magento maintenance:allow-ips --add 192.168.46.0/24 fd12:3456:789a:1::1/64`


**Before:** :heavy_multiplication_x: Not allowing
Screenshot 2023-08-30 at 12 33 16 PM
Screenshot 2023-08-30 at 12 33 23 PM


**After:** :heavy_check_mark:  Now allowing IP's

Screenshot 2023-08-31 at 12 34 10 PM

Builds are failed. Hence, moving this PR to **Extended Testing.**

**Comment by engcom-Lima** [comment]
@magento create issue

**Comment by fredden** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
Hi Fredden

Please don't run the builds. GitHub is having an incident with webhooks and delivery might be delayed, Please, do not post several comments to run the builds, they will start eventually.

**Comment by fredden** [comment]
@engcom-Lima I only ask for tests to run after pushing to this branch. (Typically this is to resolve merge conflicts, again.) It looks like the tests for this pull request don't run; there are no corresponding comments from the Adobe robot saying that the request has been received and tests queued. Please can you investigate this from the Adobe side.

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
Hi @fredden ,

Please resolve the conflicts.

Thanks

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Delta** [comment]
Hi @fredden, Thank you for your contribution. We have tested this PR again today and below error is appearing on terminal again and I have tried with the solution"composer install" and the same error is coming again.

The error

Screenshot 2024-01-30 at 8 38 20 PM

The solution i have tried

https://github.com/magento/magento2/pull/30699#issuecomment-1699377267

It would be great if you help us with this issue.

**Comment by fredden** [comment]
It looks like the dependency that I chose to perform the CIDR lookup/logic is not being maintained. I will investigate another solution.

**Review by engcom-Delta — CHANGES_REQUESTED** [review]
Earlier error is appearing on terminal again and there is no workaround as of now ,moving this to needs update until then

**Comment by fredden** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
As we are unable to run the tests on this PR we are moving this PR `on hold`.

The internal team is working on the same.

Thanks

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/App/Test/Unit/MaintenanceModeTest.php
Comment (engcom-Hotel) [inline lib/internal/Magento/Framework/App/Test/Unit/MaintenanceModeTest.php]: Please add the return type for the newly added and modified methods

```suggestion
    public function testSetSingleAddressV4Legacy(): void
```

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @fredden,

Thanks for the collaboration!

Please have a look into the review point below and resolve the failed automated tests, they seem due to PR changes.

Thanks

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Integration Tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/App/MaintenanceMode.php
Comment (engcom-Hotel) [inline lib/internal/Magento/Framework/App/MaintenanceMode.php:57]: Have you removed the question mark(?) intentionally from the `Manager`?

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, WebAPI Tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @fredden,

Thanks for the contribution!

Please have a look at the review points below and fix the failed automated tests.

Thanks

File: lib/internal/Magento/Framework/App/MaintenanceMode.php
Comment (fredden) [reply lib/internal/Magento/Framework/App/MaintenanceMode.php:57]: Yes. The value should either be supplied with the correct type, or not supplied. By specifying a default value of `= null`, the type implicitly accepts a null value. This is a documentation improvement, and not a code/policy change.

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @fredden,

There are still some automated tests failures due to code changes. Please fix them.

Thanks

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Delta** [comment]
Hi @fredden, Thanks for your contribution, We have tested in PR again and we are getting the below error while reindexing. 
Kindly suggest

Screenshot 2024-02-28 at 8 13 38 PM

**Comment by engcom-Delta** [comment]
### :heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento **`2.4-develop`**

**Manual testing scenario:**

1.Enable maintenance mode with php bin/magento maintenance:enable
2.Allow individual IP address, eg php bin/magento maintenance:allow-ips --add 192.168.10.180
3.Allow multiple individual IP addresses, eg php bin/magento maintenance:allow-ips --add 192.168.10.170 192.168.10.190
4.Allow range of IP addresses, eg php bin/magento maintenance:allow-ips --add 192.168.10.160/27
5.Allow IPv4 & IPv6 addresses, eg php bin/magento maintenance:allow-ips --add 192.168.46.0/24 fd12:3456:789a:1::1/64

**Before:** :heavy_multiplication_x: 

Screenshot 2024-02-29 at 7 13 10 PM


**After:** :heavy_check_mark:  

Screenshot 2024-02-29 at 6 16 51 PM


Builds are failed. Hence, moving this PR to **Extended Testing.**

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests

**Comment by engcom-Charlie** [comment]
Functional EE test failure are different in recent 2 builds, seems to be flaky.

**Run 1:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/30699/e74363b5e79d7df56023597bba1b86c0/Functional/allure-report-ee/index.html#categories
image


**Run 2:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/30699/1f4008749922aeb01ee8bea0e89903ca/Functional/allure-report-ee/index.html
image

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
Functional B2B failures are different in recent 2 builds with same codebase. Hence moving this PR to Merge in Progres.

**Run 1:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/30699/6df23a4e865e936e013e176643636cc0/Functional/allure-report-b2b/index.html
image


**Run 2:** 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/30699/dd1c97af15a261699fa736ffbe1b13f5/Functional/allure-report-b2b/index.html#categories/9f982a07e06d2331f9bcf457876e0a53/62a632709e465b54/
image

File: lib/internal/Magento/Framework/App/MaintenanceMode.php
Comment (thomas-kl1) [reply lib/internal/Magento/Framework/App/MaintenanceMode.php:57]: Implicit nullable is deprecated in php8.4, so this change should be reverted

Comment (fredden) [reply lib/internal/Magento/Framework/App/MaintenanceMode.php:57]: @thomas-kl1 Yes. That change to PHP happened after this code was authored. When Magento adds support for PHP v8.4+, this will be one of the (likely many) changes required.

**Pull request merged**
